File: files-app/src/main/java/l/files/fs/local/MagicFileTypeDetector.java
package l.files.fs.local;

import com.google.common.net.MediaType;

import org.apache.tika.Tika;
import org.apache.tika.io.TaggedIOException;

import java.io.IOException;
import java.io.InputStream;

import static com.google.common.net.MediaType.OCTET_STREAM;
import static l.files.fs.LinkOption.FOLLOW;

final class MagicFileTypeDetector extends LocalFileTypeDetector {

    private static final class TikaHolder {
        static final Tika tika = new Tika();
    }

    public static final MagicFileTypeDetector INSTANCE = new MagicFileTypeDetector();

    private MagicFileTypeDetector() {
    }

    @Override
    protected MediaType detectRegularFile(
            LocalResource resource,
            LocalResourceStatus status) throws IOException {

        try {

            try (InputStream in = resource.openInputStream(FOLLOW)) {
                return MediaType.parse(TikaHolder.tika.detect(in));
            }

        } catch (TaggedIOException e) {
            if (e.getCause() != null) {
                throw e.getCause();
            } else {
                throw e;
            }
        } catch (IllegalArgumentException e) {
            return OCTET_STREAM;
        }
    }

}

File: files-app/src/main/kotlin/fs/local/MagicFileTypeDetector.kt
package l.files.fs.local

import org.apache.tika.Tika

import l.files.fs.ResourceStatus
import kotlin.properties.Delegates
import com.google.common.net.MediaType
import com.google.common.net.MediaType.OCTET_STREAM
import org.apache.tika.io.TaggedIOException

private object MagicFileTypeDetector : LocalFileTypeDetector() {

    private val tika: Tika by Delegates.lazy { Tika() }

    override fun detectRegularFile(status: ResourceStatus) = try {
        MediaType.parse(tika.detect(LocalPath.check(status.getPath()).file))

    } catch (e: TaggedIOException) {
        if (e.getCause() != null) {
            throw e.getCause()
        } else {
            throw e
        }
    } catch (e: IllegalArgumentException) {
        OCTET_STREAM
    }

}
